Add fancy typehints on hover

flying-sheep opened the pull request:
- [x] doesn’t hide punctuations, we need a way around it.
- [ ] CSS adding might not be optimal, see sphinx-doc/sphinx#1379
- [ ] Checking the call frame is also not elegant, but I see no other option to patch `format_annotation`, except for replacing the version of `process_docstring` that’s probably already installed as an event handler.

Take a look on [readthedocs](https://icb-scanpydoc.readthedocs-hosted.com/en/fancy-typehints/scanpydoc.elegant_typehints.html)! On hover, `{str: Any}` (setup’s return type) becomes `Dict[str, Any]`!

Fixes #3

flying-sheep commented:
Ha! this finally works. @falexwolf would you like to take a look?

falexwolf commented:
Wow! This looks very cool! :smile:

For now, I only saw the simple example based on `Dict[]`. It's awesome! Shall we test on anndata where we have this case with all the complicated type annotations for `AnnData`? Wouldn't be a drama if something unexpected happens to the docs for a couple of days (I don't expect them to go down completely)... So I'm happy to merge!

flying-sheep commented:
I added tests, you can see that things work as expected there.

I think we should merge this as-is: The version that will be visible without hovering is the exact version that we had before I moved things into this package, with the addition of the `Dict` thing.

After we merge this, I think we should show `Mapping`/`MutableMapping`/`Dict` as `{K: V}` in the abridged version and add a demo page to http://icb-scanpydoc.readthedocs-hosted.com/

speaking of which: how did you set the custom domain here? https://readthedocs.com/dashboard/icb-scanpy/domains/

When I go to [that page for scanpydoc](https://readthedocs.com/dashboard/icb-scanpydoc/domains/), I see:

![grafik](https://user-images.githubusercontent.com/291575/51072892-2d5da600-1669-11e9-8b20-3f30e44cfe9c.png)
